### PR TITLE
✨ Date Time Library

### PR DESCRIPTION
**Pull request by Atarpara**
## Description

Added Date Time Library #152 

## Checklist

Ensure you completed **all of the steps** below before submitting your pull request:

- [x] Ran `forge snapshot`?
- [x] Ran `npm run lint`?
- [x] Ran `forge test`?

_Pull requests with an incomplete checklist will be thrown out._

### CODE REVIEW

**Comment by Atarpara** [comment]
@RogerPodacter please check this lib maybe i missed some functions please create issue on https://github.com/Atarpara/DateTimeLib.

**Comment by Vectorized** [comment]
Wow, looks good! The constant should be marked as internal.

**Comment by Atarpara** [comment]
@Vectorized If I marked internal then need to sload in assembly and gas cost increase by 70-80%.

**Comment by RogerPodacter** [comment]
Awesome!!

My quick thought is that it would be nice to have some additional helpers to make it easier to find the number of days / weeks etc between two timestamps and also to be able to find things like "the start of next week" and "the date of the first day of next year."

Even things like `addHours` in the original, which are quite simple, I believe are helpful to users. And the months and years variants are not simple!

Basically the dream would be if this were a full drop-in replacement for Bokky which then people could build on knowing it was the new meme lib to use.

Also, not sure the style guide here, but as a user I'd prefer the default to be that undefined behavior reverts().

Love this!

**Comment by Vectorized** [comment]
I think we can do reverts for undefined cases.

**Comment by Atarpara** [comment]
@RogerPodacter @Vectorized thanks for your feedback. I can start work on this.

**Comment by Atarpara** [comment]
@Vectorized @RogerPodacter I have some question about `submonths()` and `subyears()` functionality. 
Example :- 
Current date is 06-11-2022  if user want to subtract 1 month or 1 year from this date than how many days should be minus.

**Comment by Vectorized** [comment]
@Atarpara need to follow the algorithm in BokkyPooBahs library faithfully, as it deals with leap years.

**Comment by RogerPodacter** [comment]
> Current date is 06-11-2022 if user want to subtract 1 month or 1 year from this date than how many days should be minus.

This is a good point and maybe the function should have a different name like `sameDayIndexInPreviousMonth` (or something better) but I was thinking it should match the colloquial meaning, and when someone says "exactly a month ago" in that case they mean 05-11-2022.

Likewise with year. When someone says "a year ago to the DAY," they are referring to same month / day index in previous year (and are therefore ignoring leap years)

**Comment by Vectorized** [comment]
Disregard my comment if required. I haven’t accumulated enough Schizo energy to burst through this PR yet. Need to marinate more first.

**Comment by Vectorized** [comment]
@Atarpara @RogerPodacter 

Ok, I’ve looked deeper. And I like the design choices. Let’s not revert if the date is invalid. Dev will be responsible for that. This is for bytecode compactnesses. 

I think we don’t need convenience functions that can be easily represented by simple one liner arithmetic expressions.

**Comment by Vectorized** [comment]
@Atarpara What is the purpose of `getNthDayOfWeekInMonthOfYear`?

**Comment by Atarpara** [comment]
@Vectorized yes. For example some dev use `timestampFromDate()` via `isValidDate()` function so max timetamp possible is `type(uint256).max` so for that I consider year max limit.
